File: 207.course-schedule.cs
/*
 * @lc app=leetcode id=207 lang=csharp
 *
 * [207] Course Schedule
 */

// @lc code=start
using System.Collections.Generic;

public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites) {
        List<List<int>> G = new List<List<int>>();
        int[] degree = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            G.Add(new List<int>());
        }
        
        foreach (int[] prerequisite in prerequisites) {
            int j = prerequisite[0];
            int i = prerequisite[1];
            G[i].Add(j);
            degree[j]++;
        }
        
        Queue<int> dq = new Queue<int>();
        HashSet<int> seen = new HashSet<int>();
        for (int i = 0; i < numCourses; i++)
        {
            if (degree[i] == 0)
            {
                dq.Enqueue(i);
                seen.Add(i);
            }
        }
        
        while (dq.TryDequeue(out int i))
        {
            foreach (int j in G[i])
            {
                degree[j]--;
                if (degree[j] == 0)
                {
                    dq.Enqueue(j);
                    seen.Add(j);
                }
            }
        }
        return seen.Count == numCourses;
    }
}

// @lc code=end

File: 207.course-schedule.py
#
# @lc app=leetcode id=207 lang=python3
#
# [207] Course Schedule
#

# @lc code=start
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        G = [[] for i in range(numCourses)]
        degree = [0] * numCourses
        for j, i in prerequisites:
            G[i].append(j)
            degree[j] += 1
        bfs = [i for i in range(numCourses) if degree[i] == 0]
        for i in bfs:
            for j in G[i]:
                degree[j] -= 1
                if degree[j] == 0:
                    bfs.append(j)
        return len(bfs) == numCourses
# @lc code=end

